app/queries/get_grouped_content_and_links.rb

```diff
@@ -0,0 +1,116 @@
+
+    def initialize(last_seen_content_id: nil, page_size: nil)
+      @last_seen_content_id = last_seen_content_id || DEFAULT_CONTENT_ID
+      @page_size = page_size || PAGE_SIZE
+    end
+
+    # Returns an array of hashes where each hash contains a content_id, an
+    # array of content item rows, and an array of link rows.
+    def call
+      content_ids = query_content_ids_for_page(
+        last_seen_content_id: @last_seen_content_id,
+        page_size: @page_size
+      )
+
+      group_results(content_results(content_ids), link_set_results(content_ids))
+    end
+
+    def has_valid_uuid_for_pagination
+      unless @last_seen_content_id == DEFAULT_CONTENT_ID || UuidValidator.valid?(@last_seen_content_id)
+        errors.add(:last_seen_content_id)
+      end
+    end
+
+  private
+
+    def query_content_ids_for_page(last_seen_content_id:, page_size:)
+      ContentItem
+        .select(:content_id)
+        .group(:content_id)
+        .having("content_id > ?", [last_seen_content_id])
+        .limit(page_size)
+        .order(:content_id)
+        .map(&:content_id)
+    end
+
+    def group_results(content_results, link_set_results)
+      grouped_content_items = group_by_content_id(content_results)
+      grouped_links = group_by_content_id(link_set_results)
+
+      grouped_content_items.map do |content_id, content_items|
+        {
+          "content_id" => content_id,
+          "content_items" => content_items,
+          "links" => grouped_links[content_id] || []
+        }
+      end
+    end
+
+    def group_by_content_id(results)
+      results.group_by { |item| item["content_id"] }
+    end
+
+    def content_results(content_ids)
+      return [] if content_ids.empty?
+
+      query = <<-SQL
+        SELECT
+          ci.content_id as content_id,
+          ci.id as content_item_id,
+          t.locale as locale,
+          loc.base_path as base_path,
+          s.name as state,
+          ufv.number as user_facing_version,
+          ci.publishing_app,
+          ci.format
+
+        FROM
+          content_items ci
+        JOIN translations t on t.content_item_id = ci.id
+        JOIN locations loc on loc.content_item_id = ci.id
+        JOIN states s on s.content_item_id = ci.id
+        JOIN user_facing_versions ufv on ufv.content_item_id = ci.id
+
+        WHERE ci.content_id IN (#{sql_value_placeholders(content_ids)})
+        ORDER BY ci.updated_at DESC
+      SQL
+
+      ActiveRecord::Base.connection.raw_connection.exec(query, content_ids)
+    end
+
+    def link_set_results(content_ids)
+      return [] if content_ids.empty?
+
+      query = <<-SQL
+        SELECT
+          link_sets.content_id,
+          links.link_type,
+          links.target_content_id
+        FROM
+          link_sets
+        JOIN
+          links on link_sets.id = links.link_set_id
+
+        WHERE link_sets.content_id IN (#{sql_value_placeholders(content_ids)})
```

Comment:
`sql_value_placeholders` might never be anything dangerous right now but I think that adding an extra level of security around this query will not hurt and will prevent SQL injection in the case that in the future the code will change but the query will remain the same. 

Sadly, I appear not to find an easy way to escape SQL in multiline strings.